feat: compare reference and target SBOMs, show extra components in target

baprusty opened the pull request:
Add a `compare` command to identify components present in the target SBOM but
    missing in the reference/base SBOM. Highlight these extra components and create
    a new SBOM containing only the extras. Help track already sent or cleared
    components and identify additional components that require license clearing.

    Signed-off-by: badrikesh prusty <badrikesh.prusty@siemens.com>

fmoessbauer commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,239 @@
+            sbom (str): Path or URL to the SBOM file.
+            sbom_tmp (str): Temporary destination path to copy/download the SBOM.
+        """
+        if sbom.startswith(('http://', 'https://', 'ftp://')):
```
At least in the first version of this patchset I prefer to only support local files. By that, we can avoid network access which needs some additional considerations.

baprusty commented:
Update: I’ve migrated the comparison script to use the existing Debsbom APIs along with other improvements.
Current status: it successfully generates the extra component SBOM, but it throws some dependency-chain warnings for the CycloneDX format during generation. I’m investigating the issue and working on a fix.

baprusty commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,239 @@
+            help="Path or URL to the target (new) SBOM file"
+        )
+
+        parser.add_argument(
```
updated, thanks!

Checked with below use cases:
```
(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json) -n <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json) -t spdx
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json) -n <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json) -t cdx
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json
ERROR:debsbom.cli:can not compare mixed SPDX and CycloneDX documents
debsbom: error: can not compare mixed SPDX and CycloneDX documents


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)cat  ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json | debsbom compare -b - -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json -t cdx


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)cat  ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json | debsbom compare -b - -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json -t spdx


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json
(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json --out -
{
    "components": [
        {
            "bom-ref": "pkg:deb/debian/acl@2.3.1-3?arch=arm64",
            "description": "access control list - utilities\nThis package contains the getfacl and setfacl utilities needed for manipulating access control lists. It also contains the chacl IRIX compatible utility.",
            "externalReferences": [
                {
                    "comment": "homepage",
                    "type": "website",
                    "url": "https://savannah.nongnu.org/projects/acl/"
                }
            ],
...
...
...
```

baprusty commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,239 @@
+    @classmethod
+    def load_spdx_sbom(cls, path):
+        """Return packages keyed by purl or fallback name@version."""
+        with open(path) as f:
```
Updated, thanks. Tried using already present API instead of creating our own.

baprusty commented:
> Sorry for joining the discussion late, but I wonder, is it actually good to include such a feature in debsbom? This can get a rather involved effort, thinking of possible naming or PURL deviations, need to compare based on some extra properties (conveying workflow information) etc. Perhaps it would be better to instead defer this to other tools already providing such functionality for cdx or spdx?
> 
> Sorry if I overlooked it, but is there some description about the goals how exactly the comparison shall work with limitations?

Hi @gernot-h , We tried existing tools like CycloneDX CLI (https://github.com/CycloneDX/cyclonedx-cli), but the output was not accurate. For example:
- It only considers a component identical if all fields match 100%. When comparing SBOMs, one SBOM had `"group": "debian"` for each component, while the other SBOM was missing this field, causing none of the components to match.
- The output generated by `cyclonedx-cli diff` creates a large file consisting of components of both SBOMs in a format like:
```
{
  "componentVersions": {
    "adduser": {
      "added": [],
      "removed": [],
      "unchanged": [ <component-details> ]
        },
```

baprusty commented:
MR is ready for review. Please take a look, thanks!

Tested with below use cases:
```
(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json) -n <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json) -t spdx
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json) -n <(cat ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json) -t cdx
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.
WARNING:debsbom.commands.input:Expecting data via stdin, but connected to TTY.


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json
ERROR:debsbom.cli:can not compare mixed SPDX and CycloneDX documents
debsbom: error: can not compare mixed SPDX and CycloneDX documents


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)cat  ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json | debsbom compare -b - -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json -t cdx


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)cat  ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json | debsbom compare -b - -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json -t spdx


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.spdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.spdx.json
(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json


(python-venv) bp@ipc847e:~/test12/debsbom2$ (bp/compare-sboms)debsbom compare -b ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LX1_sbom/sbom.cdx.json -n ~/LPE_SBOM_COMPARISON_WITH_LX1_SUBMITTED_SBOM/LPE_submitted_SBOM/audis-lpe-image-release-audis-lpe-lpe9403-std.cdx.json --out -
{
    "components": [
        {
            "bom-ref": "pkg:deb/debian/acl@2.3.1-3?arch=arm64",
            "description": "access control list - utilities\nThis package contains the getfacl and setfacl utilities needed for manipulating access control lists. It also contains the chacl IRIX compatible utility.",
            "externalReferences": [
                {
                    "comment": "homepage",
                    "type": "website",
                    "url": "https://savannah.nongnu.org/projects/acl/"
                }
            ],
...
...
...
```

fmoessbauer commented on docs/source/commands/compare.rst:
```diff
@@ -0,0 +1,10 @@
+``compare`` command
+==================
+
+.. automodule:: debsbom.commands.compare.CompareCmd
```
This needs to be aligned with the manpage addition in #155 , but we can take care of that.

fmoessbauer commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,169 @@
+
+        for kind, sbom_arg in inputs:
+            # read from stdin
+            if sbom_arg == "-" or sbom_arg.startswith("/dev/fd/"):
```
Why `/dev/fd/`? The other commands also only allow stdin and not arbitrary file descriptors.

fmoessbauer commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,169 @@
+            "-t",
+            "--sbom-type",
+            choices=["cdx", "spdx"],
+            help="expected SBOM type when reading SBOMs from stdin, required when reading from stdin",
```
@Urist-McGit Do we already have this as generic argument? By that, we could avoid description string drift for the same thing across commands.

baprusty replied:
Here, I am using only `GenerateInput` and due to this `sbom-type` is not available. 

The `SbomInput` brings in additional `bomin` args as well and we are not using it. currently bomin can only take one argument.
May be later we can refactor it to take multiple SBOMs arguments or add a check to skip bomin using if not required.

fmoessbauer reviewed (commented):
The change itself is fine for me. While I get @gernot-h 's arguments, I also had trouble comparing the SBOMs with generic tooling. For simple cases, that works, but for more complex ones we IMHO need tooling support - as introduced here.

Apart from the required code changes, we MUST have tests for this feature.

baprusty commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,172 @@
+
+        for kind, sbom_arg in inputs:
+            # read from stdin
+            if sbom_arg == "-":
```
Thank you for the review and the great suggestion. I've updated it accordingly. Thanks again!

baprusty commented on src/debsbom/commands/compare.py:
```diff
@@ -0,0 +1,169 @@
+                if args.sbom_type is None:
+                    raise ValueError("option --sbom-type is required when reading SBOMs from stdin")
+                # determine file descriptor
+                if sbom_arg == "-":
```
I searched but couldn’t find an existing wrapper for this case.

gernot-h commented:
> Sorry for the long silence, pretty busy with workshops these days. If there's still time, I'd like to give the new feature a try with some of my use cases in the next days and report back.

So I tested some simple cases and the new command seems to work as expected. I'd have loved to try more real-world scenarios, but unfortunately, the service I wanted to extract and compare some SBOMs from is currently down and I've no clue if this will still be fixed before Christmas, so don't want to block this any longer.

fmoessbauer commented:
Hi @baprusty , I just reworked this PR on top of the recent debsbom changes in #170 . While doing so, I notices the following things that need to be fixed before this MR can be merged:

- generated SBOMs are semantically incorrect (test with SPDX and `--validate`), as they contain references to components that are not in the SBOM
- the naming `compare` is confusing: For a compare, I expect to get a overview of what changed, but not another SBOM. I would rather call the command `delta` or `sbom-diff`.
- You removed the `cyclonedx-cli diff` and `diff / jq` examples to get a list of changed components. With this MR, there is no such option to generate this list, so please keep the example as-is.
- tests are missing

When continuing the work, feel free to use #170 as a base.

baprusty commented:
> Hi @baprusty , I just reworked this PR on top of the recent debsbom changes in #170 . While doing so, I notices the following things that need to be fixed before this MR can be merged:
> 
> * generated SBOMs are semantically incorrect (test with SPDX and `--validate`), as they contain references to components that are not in the SBOM
> * the naming `compare` is confusing: For a compare, I expect to get a overview of what changed, but not another SBOM. I would rather call the command `delta` or `sbom-diff`.
> * You removed the `cyclonedx-cli diff` and `diff / jq` examples to get a list of changed components. With this MR, there is no such option to generate this list, so please keep the example as-is.
> * tests are missing
> 
> When continuing the work, feel free to use #170 as a base.

Sure Felix, thanks for feedback, will work on it and update.

baprusty commented:
> Hi @baprusty , I just reworked this PR on top of the recent debsbom changes in #170 . While doing so, I notices the following things that need to be fixed before this MR can be merged:
> 
> * generated SBOMs are semantically incorrect (test with SPDX and `--validate`), as they contain references to components that are not in the SBOM
> * the naming `compare` is confusing: For a compare, I expect to get a overview of what changed, but not another SBOM. I would rather call the command `delta` or `sbom-diff`.
> * You removed the `cyclonedx-cli diff` and `diff / jq` examples to get a list of changed components. With this MR, there is no such option to generate this list, so please keep the example as-is.
> * tests are missing
> 
> When continuing the work, feel free to use #170 as a base.

Hi @fmoessbauer , thanks for rework with latest changes. I pulled in your changes from #170 to this MR as I don't have permissions to push to source repo directly.

Addressed the comments:
1. Appended distro_relationship to relationships to fix missing components and added the related SPDX element ID only if it is present in id_map.
2. Updated `debsbom compare` to `debsbom delta` command.
3. Added back cyclonedx-cli diff and diff / jq examples to get a list of changed components, and created separate entry for delta command.
4. Added CI tests and pytest scripts for delta command.

For some reason, CI complained about untouched files due to 2 empty lines between imports and logger definition. Link: https://github.com/siemens/debsbom/actions/runs/21119056371/job/60728967004?pr=121
Kept one empty line between import and logger definition to resolve the issue.

Please review.

fmoessbauer commented:
Hi @baprusty , the implementation in general is fine now. Thanks! However, I'm still able to get validation errors. To easily reproduce, take two sboms from different distributions (e.g. bookworm and trixie), generate the SPDX and run with `--validate`.

fmoessbauer commented:
Thanks for the patience. Merged.
